packages/plugins/plugin-table/src/model/modal-controller.ts

```diff
@@ -0,0 +1,83 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { signal } from '@preact/signals-core';
+import { type RefObject, type MouseEvent, type MutableRefObject } from 'react';
+
+import { tableButtons } from '../util';
+
+export type ColumnSettingsMode = { type: 'create' } | { type: 'edit'; fieldId: string };
+
+export type ModalState =
+  | { type: 'column'; fieldId: string }
+  | { type: 'row'; rowIndex: number }
+  | { type: 'columnSettings'; mode: ColumnSettingsMode }
+  | { type: 'closed' };
+
+export class ModalController {
+  private readonly _state = signal<ModalState>({ type: 'closed' });
+  private readonly _triggerRef: MutableRefObject<HTMLElement> = { current: null as unknown as HTMLElement };
+
+  public get state() {
+    return this._state;
+  }
+
+  public get trigger() {
+    return this._triggerRef as RefObject<HTMLButtonElement>;
+  }
+
+  public handleClick = (event: MouseEvent): boolean => {
+    const target = event.target as HTMLElement;
+
+    const columnButton = target.closest(`button[${tableButtons.columnSettings.attr}]`);
```

Comment:
perhaps keep these in order `row, column, columnSettings, close` and match the variable name to the state  (`newColumnButton => columnSettingsButton`)